### PR TITLE
Bigc 679 update cart error handling

### PR DESCRIPTION
**Pull request by sofib**

### CODE REVIEW

File: extension/lib/shopgate/Concurrency.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+const { sleep } = require('./concurrency/util')
+const { ShopgateConcurrencyLock, useStorage, validLockExists } = require('./concurrency/Lock')
+
+class ShopgateConcurrency {
```
Comment (chrisEff) [inline extension/lib/shopgate/Concurrency.js]: this class has no state at all. all its methods could be static. but then again, why do we need a class at all? could as well just export a bunch of functions...

Comment (sofib) [reply extension/lib/shopgate/Concurrency.js]: to me it looked nice from the perspective of a caller/consumer of the code, and I do like to have an entry point to the feature/domain as it helps in dealing with dependencies

File: extension/lib/shopgate/concurrency/util.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ */
+function sleep (ms) {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+/**
+ * @param {number} [length=32]
+ * @return {string}
+ */
+function randomString (length = 32) {
```
Comment (chrisEff) [inline extension/lib/shopgate/concurrency/util.js]: randomString is unused

File: extension/lib/shopgate/CartExtensionPipeline.js [hunk 3]
```diff
@@ -98,21 +123,51 @@ class ShopgateCartExtensionPipeline {
...
+          return BigCommerceCartRepository.createLineItemUpdate(item.CartItemId, item.quantity)
+        }),
+        (failureEvent) => {
+          this._context.log.error(decorateDebug({
+            reason: failureEvent.reason,
+            cartItemId: failureEvent.item.itemId,
+            quantity: failureEvent.item.quantity
+          }), 'Failed updating product')
+
+          this._messagesRepository.push(cartId, failureEvent.reason, failureEvent.item.itemId)
```
Comment (chrisEff) [inline extension/lib/shopgate/CartExtensionPipeline.js]: I'm not sure, does this need to be awaited?

File: extension/lib/shopgate/concurrency/Lock.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+/**
+ * Determines if a lock was successful.
+ *
+ * @param {ShopgateConcurrencyLock} lockInstance
+ * @param {number} maxWaitTime
+ * @returns {boolean}
+ */
+async function elect (lockInstance, maxWaitTime) {
```
[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/shopgate/concurrency/Lock.js]: sounds nicer than chose ;)

[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/shopgate/concurrency/Lock.js]: I didn't want to have it exposed, this way it is private

File: extension/lib/shopgate/CartExtensionPipeline.js (hunk 3 above)
Comment (sofib) [reply extension/lib/shopgate/CartExtensionPipeline.js]: let's talk

File: extension/lib/shopgate/concurrency/util.js (hunk 2 above)
Comment (sofib) [reply extension/lib/shopgate/concurrency/util.js]: removed

File: extension/lib/shopgate/concurrency/Lock.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+class ShopgateConcurrencyLock {
+  /**
+   * @param {string} name
+   * @param {string} owner
+   * @param {number} ttl Time to live in milliseconds
+   */
+  constructor (name, owner, ttl) {
+    this._name = name
+    this._owner = owner
+    this._ttl = ttl
```
[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/shopgate/concurrency/Lock.js]: I kinda felt it belong there even though it has no usage, I've removed it

File: extension/lib/shopgate/concurrency/Lock.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisEff) [reply extension/lib/shopgate/concurrency/Lock.js]: I'd say let's use a combination of @private and underscore. That's what horizon does in the SDK. So let's be consistent with that.

[the remark this answers is not in the document]
Comment (chrisEff) [reply extension/lib/shopgate/concurrency/Lock.js]: like discussed, chose also sounds a bit odd...

File: extension/lib/shopgate/Concurrency.js (hunk 1 above)
Comment (chrisEff) [reply extension/lib/shopgate/Concurrency.js]: hmm, well i'm not fully convinced, but ok

File: extension/lib/shopgate/CartExtensionPipeline.js (hunk 3 above)
Comment (chrisEff) [reply extension/lib/shopgate/CartExtensionPipeline.js]: I guess I'd say let's await.

File: extension/lib/shopgate/concurrency/Lock.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/shopgate/concurrency/Lock.js]: if you want it changed, would be nice to suggest something as well

[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/shopgate/concurrency/Lock.js]: I think we'll have to learn to live with Javascript and how things can really be private there where appropriate. But for the sake of completing this, I don't wont to discuss further - done

File: extension/lib/shopgate/CartExtensionPipeline.js (hunk 3 above)
Comment (sofib) [reply extension/lib/shopgate/CartExtensionPipeline.js]: done

File: extension/lib/shopgate/concurrency/Lock.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/shopgate/concurrency/Lock.js]: I've renamed it, if you still don't like ti please make a suggestion

**Pull request merged**